crates/bevy_ecs/src/storage/table.rs

```diff
@@ -101,15 +102,19 @@ impl TableId {
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
 // SAFETY: Must be repr(transparent) due to the safety requirements on EntityLocation
 #[repr(transparent)]
-pub struct TableRow(u32);
+pub struct TableRow(NonMaxU32);
 
 impl TableRow {
-    pub(crate) const INVALID: TableRow = TableRow(u32::MAX);
+    pub(crate) const INVALID: TableRow = TableRow(NonMaxU32::MAX);
 
     /// Creates a `TableRow`.
+    ///
+    /// Will panic if the provided value does not fit within a [`NonMaxU32`].
     #[inline]
     pub const fn from_u32(index: u32) -> Self {
-        Self(index)
+        debug_assert!(index <= NonMaxU32::MAX.get());
+        // SAFETY: Caller guarantees that `index` is valid
```

Comment:
This comment is wrong, and the assertion that's supposed to guarantee this is a `debug_assert!` instead of a normal `assert!`, so that wouldn't be sound

Reply:
yeah ,In theory it is unsound , but in practice there should not exist more than 4 billions of entities, Tablerow is used in many performance sensitive areas where any branch maybe unacceptable

Reply:
`TableRow` is a public API, so this makes it trivial to cause UB from safe code that uses bevy. Maybe add an unsafe `from_u32_unchecked` for internal use, if the check is really so devastating to performance?

Reply:
what's your opions about If we don't change the `TableRow `and directly using `NonmaxU32 `on the ComponentSparseSet::SparseArray ?  IMO it looks like a more safer way  as we don't change any user side API and u32:: max happens to corresponds to TableRow:: INVAlID

Reply:
I'll let the ECS maintainers weigh in on that, since I don't really know the internals of that code well. To me it looks like either way would work, but having the niche in `TableRow` may benefit external users of that type (but it doesn't look like that's a terribly common use case?).

Reply:
I'm fine with that approach: I think it's a bit less intrusive.